tangelo/linq/circuit.py

```diff
@@ -415,17 +443,80 @@ def remove_small_rotations(circuit, param_threshold=0.05):
     rot_gates = {"RX", "RY", "RZ", "CRX", "CRY", "CRZ"}
     gates = [g for g in circuit._gates if not (g.name in rot_gates and abs(g.parameter) % (2*np.pi) < param_threshold)]
 
-    return Circuit(gates)
+    return Circuit(gates) if remove_qubits else Circuit(gates, n_qubits=circuit.width)
+
+
+def merge_rotations(circuit: Circuit):
+    """ Merge rotation gates successively applied to the same qubits (targets, controls), into a single
+    equivalent rotation gate. If one of the gates involved is labeled variational, the merged rotation is
+    as well.
+
+    Agglomerating non-numerical values for parameters is currently unsupported.
+
+    Args:
+        circuit (Circuit): Input quantum circuit
+
+    Returns:
+        Circuit: circuit with merged rotations
+    """
+
+    NoneGate = Gate('NONE', 0)
+
+    gate_qubits = {i: list() for i in range(circuit.width)}
+    new_gates = []
+
+    rot_gates = {"RX", "RY", "RZ", "CRX", "CRY", "CRZ"}
+
+    for gi, gate in enumerate(circuit):
+        merge_gate = False
+
+        # Identify qubits the current gate acts on.
+        qubits = gate.target if gate.control is None else gate.target + gate.control
+
+        # For all qubits involved, inspect the previous gate.
+        g_prevs = [gate_qubits[qubit_i][-1][1] if gate_qubits[qubit_i] else NoneGate for qubit_i in qubits]
+
+        # No rotations merged: no single previous gate found operating on all qubits involved.
+        if (NoneGate in g_prevs):
```

Comment:
Why creating a `NoneGate` object, `None` would not be enough?

Reply:
It was the first thing I tried and made the code more complicated, as generic python control flow turned out to implicitly use the `==` comparison between objects, which did not work between `None` and a `Gate` object.

Maybe I could have been smarter.

Reply:
Ah I see the problem (the `in` keyword forces a comparison on all objects)! Good to know!